ledger/src/shred.rs

```diff
@@ -653,48 +662,95 @@ impl TryFrom<u8> for ShredVariant {
     }
 }
 
-// Get slot, index, and type from a packet with partial deserialize
-pub fn get_shred_slot_index_type(
+#[must_use]
+pub fn should_discard_shred(
     packet: &Packet,
+    root: Slot,
+    shred_version: u16,
+    // Range of slots to ingest shreds for.
+    slot_bounds: impl RangeBounds<Slot>,
     stats: &mut ShredFetchStats,
-) -> Option<(Slot, u32, ShredType)> {
+) -> bool {
     let shred = match layout::get_shred(packet) {
         None => {
             stats.index_overrun += 1;
-            return None;
+            return true;
         }
         Some(shred) => shred,
     };
     if OFFSET_OF_SHRED_INDEX + SIZE_OF_SHRED_INDEX > shred.len() {
         stats.index_overrun += 1;
-        return None;
+        return true;
     }
     let shred_type = match layout::get_shred_type(shred) {
         Ok(shred_type) => shred_type,
         Err(_) => {
             stats.bad_shred_type += 1;
-            return None;
+            return true;
         }
     };
     let slot = match layout::get_slot(shred) {
-        Some(slot) => slot,
+        Some(slot) => {
+            if slot <= root || !slot_bounds.contains(&slot) {
+                stats.slot_out_of_range += 1;
+                return true;
+            }
+            slot
+        }
         None => {
             stats.slot_bad_deserialize += 1;
-            return None;
+            return true;
         }
     };
     let index = match layout::get_index(shred) {
         Some(index) => index,
         None => {
             stats.index_bad_deserialize += 1;
-            return None;
+            return true;
         }
     };
-    if index >= MAX_DATA_SHREDS_PER_SLOT as u32 {
-        stats.index_out_of_bounds += 1;
-        return None;
+    if layout::get_version(shred) != Some(shred_version) {
+        stats.shred_version_mismatch += 1;
+        return true;
+    }
+    match shred_type {
+        ShredType::Code => {
+            if index >= shred_code::MAX_CODE_SHREDS_PER_SLOT as u32 {
+                stats.index_out_of_bounds += 1;
+                return true;
+            }
+            false
+        }
+        ShredType::Data => {
+            if index >= MAX_DATA_SHREDS_PER_SLOT as u32 {
+                stats.index_out_of_bounds += 1;
+                return true;
+            }
+            let parent_offset = match layout::get_parent_offset(shred) {
+                Some(parent_offset) => parent_offset,
+                None => {
+                    stats.bad_parent_offset += 1;
+                    return true;
+                }
+            };
+            if parent_offset == 0 && slot != 0 {
+                stats.bad_parent_offset += 1;
+                return true;
+            }
```

Comment:
We have some slightly duplicated checking happening on data shreds:
1) We error if `slot <= root` on line 694 (all shreds)
2) We later error `if parent_offset == 0 && slot != 0` on line 736 (data shreds only)
3) The comparisons from 1. and 2., plus some extras are checked in `blockstore::verify_shred_slots()` on line 747 (data shreds only)
```
    if slot == 0 && parent == 0 && root == 0 {
        return true; // valid write to slot zero.
    }
    // Ignore shreds that chain to slots before the root,
    // or have invalid parent >= slot.
    root <= parent && parent < slot
```
So, I think we could:
- Drop the check in 2.
- Move the check to ensure `slot > root` from 694 to only happen on coding shreds, as `verify_shred_slots()` will confirm `slot > root` for data shreds

I can throw up a small PR for this later

Reply:
@steviez yeah, that sounds right, thanks